src/common/internet.rs

```diff
@@ -195,23 +206,35 @@ impl From<OnionAddressV3> for InetAddr {
     }
 }
 
+#[cfg(feature = "tor")]
+impl From<OnionAddressV2> for InetAddr {
+    #[inline]
+    fn from(addr: OnionAddressV2) -> Self {
+        InetAddr::TorV2(addr)
+    }
+}
+
 impl_try_from_stringly_standard!(InetAddr);
 impl_into_stringly_standard!(InetAddr);
 
 impl FromStr for InetAddr {
     type Err = String;
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        match IpAddr::from_str(s) {
-            Ok(ip_addr) => Ok(Self::from(ip_addr)),
+        match (
+            IpAddr::from_str(s),
+            OnionAddressV3::from_str(s),
+            OnionAddressV2::from_str(s),
+        ) {
+            (Ok(ip_addr), _, _) => Ok(Self::from(ip_addr)),
             #[cfg(feature = "tor")]
-            Err(_) => Ok(Self::from(
-                OnionAddressV3::from_str(s)
-                    .map(Self::from)
-                    .map_err(|_| String::from("Wrong onion address string"))?,
-            )),
```

Comment:
I'd suggest adding something like:
`(Ok(_), Ok(_), _) | (Ok(_), _, Ok(_)) | (_, Ok(_), Ok(_)) => panic!("Confusing result of parsing {}", s),`

To catch possible problems.

Reply:
Good point!